modules/sdk/src/main/scala/org/tessellation/sdk/infrastructure/consensus/ConsensusHandler.scala

```diff
@@ -21,43 +24,53 @@ object ConsensusHandler {
 
   def make[F[_]: Async: KryoSerializer: SecurityProvider, Event <: AnyRef: TypeTag: ClassTag, Key: Show: TypeTag: ClassTag, Artifact <: AnyRef: TypeTag](
     storage: ConsensusStorage[F, Event, Key, Artifact],
-    manager: ConsensusManager[F, Event, Key, Artifact]
+    manager: ConsensusManager[F, Event, Key, Artifact],
+    whitelisting: Option[Set[PeerId]]
   ): RumorHandler[F] = {
 
     val logger = Slf4jLogger.getLogger[F]
 
-    val eventHandler = RumorHandler.fromPeerRumorConsumer[F, ConsensusEvent[Event]]() { rumor =>
-      storage.addEvent(rumor.origin, (rumor.ordinal, rumor.content.value)) >>
-        manager.checkForTrigger(rumor.content.value)
-    }
+    def filterWhitelisted[A](fn: PeerRumor[A] => F[Unit]): PeerRumor[A] => F[Unit] =
+      rumor =>
+        if (whitelisting.forall(_.contains(rumor.origin)))
+          fn(rumor)
+        else
+          logger.warn(s"Non whitelisted peer ${rumor.origin} cannot participate in consensus")
 
-    val facilityHandler = RumorHandler.fromPeerRumorConsumer[F, ConsensusFacility[Key]]() { rumor =>
-      manager.addFacilitators(rumor.content.key, rumor.content.facilitators) >>
-        storage.addFacility(rumor.origin, rumor.content.key, rumor.content.bound) >>=
-        manager.checkForStateUpdate(
-          rumor.content.key
-        )
+    val eventHandler = RumorHandler.fromPeerRumorConsumer[F, ConsensusEvent[Event]]() {
+      filterWhitelisted { rumor =>
+        storage.addEvent(rumor.origin, (rumor.ordinal, rumor.content.value)) >>
+          manager.checkForTrigger(rumor.content.value)
+      }
     }
 
-    val proposalHandler = RumorHandler.fromPeerRumorConsumer[F, ConsensusProposal[Key]]() { rumor =>
-      storage.addProposal(rumor.origin, rumor.content.key, rumor.content.hash) >>= manager
-        .checkForStateUpdate(
-          rumor.content.key
-        )
+    val facilityHandler =
+      RumorHandler.fromPeerRumorConsumer[F, ConsensusPeerDeclaration[Key, Facility]]() {
+        filterWhitelisted { rumor =>
+          storage.addFacility(rumor.origin, rumor.content.key, rumor.content.declaration) >>=
+            manager.checkForStateUpdate(rumor.content.key)
+        }
+      }
+
+    val proposalHandler = RumorHandler.fromPeerRumorConsumer[F, ConsensusPeerDeclaration[Key, Proposal]]() {
+      filterWhitelisted { rumor =>
+        storage.addProposal(rumor.origin, rumor.content.key, rumor.content.declaration) >>=
+          manager.checkForStateUpdate(rumor.content.key)
+      }
     }
 
     val artifactHandler = RumorHandler.fromCommonRumorConsumer[F, ConsensusArtifact[Key, Artifact]] { rumor =>
-      storage.addArtifact(rumor.content.key, rumor.content.artifact) >>= manager.checkForStateUpdate(
-        rumor.content.key
-      )
+      storage.addArtifact(rumor.content.key, rumor.content.artifact) >>=
```

Comment:
Artifact spreads as a common rumor, so it doesn't have a single author to which we could apply the whitelisting filter. It won't affect the logic - for further processing, we're still going to use the artifact that was picked as a majority. It could however lead to memory overuse if there were malicious nodes spamming artifacts. I don't have a good idea how to solve it yet.